pyacq/core/stream.py

```diff
@@ -523,11 +576,348 @@ def get_array_slice(self, index, length, autoswapaxes = True):
+        # Note: read_index and write_index are defined independently to avoid
+        # race condifions with processes reading and writing from the same
+        # shared memory simultaneously. When new data arrives:
+        #   1. write_index is increased to indicate that the buffer has advanced
+        #      and some old data is no longer valid
+        #   2. new data is written over the old buffer data
+        #   3. read_index is increased to indicate that the new data is now
+        #      readable
+
+        #
+        #              write_index+1-bsize   break_index     read_index       write_index
+        #              |                     |               |                |
+        #    ..........<.....................|...............>................v
+        #                                    |
+        #              [           readable area             ][ writable area ]
+        #                                    |
+        #                                    |  [........]           read without copy
+        #                        [........]  |                       read without copy
+        #                               [....|......]                read with copy
+        # 
+        
+        shape = (shape[0] * (2 if double else 1),) + shape[1:]
+        # initialize int buffers with 0 and float buffers with nan
+        if fill is None:
+            fill = 0 if np.dtype(dtype).kind in 'ui' else np.nan
+        self._filler = fill
+        
+        self.buffer = np.empty(shape, dtype=dtype)
+        self.buffer[:] = self._filler
+        self.dtype = self.buffer.dtype
+        
+    def last_index(self):
+        return self._read_index
+
+    def first_index(self):
+        return self._read_index + 1 - self.shape[0]
+
+    def new_chunk(self, data, index=None):
+        dsize = data.shape[0]
+        bsize = self.shape[0]
+        if dsize > bsize:
+            raise ValueError("Data chunk size %d is too large for ring "
+                            "buffer of size %d." % (dsize, bsize))
+        if data.dtype != self.dtype:
+            raise TypeError("Data has incorrect dtype %s (buffer requires %s)" %
+                            (data.dtype, self.dtype))
+        
+        # by default, index advances by the size of the chunk
+        if index is None:
+            index = self._write_index + dsize
+        
+        assert dsize <= index - self._write_index, ("Data size is %d, but index "
+                                                    "only advanced by %d." % 
+                                                    (dsize, index-self._write_index)) 
+
+        revert_inds = [self._read_index, self._write_index]
+        try:
+            # advance write index. This immediately prevents other processes from
+            # accessing memory that is about to be overwritten.
+            self._set_write_index(index)
+            
+            # decide if any skipped data needs to be filled in
```

Comment:
normally if (index is None) next 5 lines areunessecery no ?
Am I wrong ?
So maybe we could add if index is None: 
This would avoid these lines but need an extra variable _index = index some lines before.
This function will be called heavlly. We must carefull about all checks.